File: src/StickedWords.Domain/Models/Exercises/TranslateGuessResult.cs
namespace StickedWords.Domain.Models.Exercises;

public record TranslateGuessResult : IExerciseResult
{
    private TranslateGuessResult() { }

    public GuessResult Result { get; init; }

    public string? CorrectTranslation { get; init; }

    public required bool IsExpired { get; init; }

    public static TranslateGuessResult Correct(LearningSessionState sessionState)
    {
        return new()
        {
            Result = GuessResult.Correct,
            IsExpired = sessionState is LearningSessionState.Expired
        };
    }

    public static TranslateGuessResult Wrong(FlashCardWord translation, LearningSessionState sessionState)
    {
        return new()
        {
            Result = GuessResult.Wrong,
            CorrectTranslation = translation.Word,
            IsExpired = sessionState is LearningSessionState.Expired
        };
    }
}

File: src/StickedWords.UI.React/src/models/exercises/TranslateGuessResult.ts
export enum GuessResult {
  None = "None",
  Correct = "Correct",
  Wrong = "Wrong"
}

export class TranslateGuessResult {
  
  result!: GuessResult;
  correctTranslation?: string;
  isExpired!: boolean;

  static fromJson(json: any): TranslateGuessResult {
    const result = new TranslateGuessResult();

    result.result = TranslateGuessResult.mapGuessResult(json.result);
    result.correctTranslation = json.correctTranslation ?? null;
    result.isExpired = json.isExpired ?? false;

    return result;
  }

  private static mapGuessResult(src: string): GuessResult {
    if (!src) {
      return GuessResult.None;
    }

    return Object.values(GuessResult).includes(src as GuessResult)
      ? src as GuessResult
      : GuessResult.None;
  }
}
